app/controllers/concerns/authentication.rb

```diff
@@ -96,43 +157,42 @@ def query_token_authenticator(query_token)
     end
   end
 
-  def http_password_authenticator(username = nil, password = nil)
+  def http_password_authenticator(email = nil, password = nil)
+    raise Keygen::Error::InvalidCredentialsError.new(header: 'Authorization') if
+      email.blank?
+
+    if current_account.sso? && current_account.sso_for?(email) # for JIT-user-provisioning
+      redirect = sso_authorization_url_for(email)
+
+      raise Keygen::Error::SingleSignOnRequiredError.new(links: { redirect: })
+    end
+
     user = current_account.users.for_environment(current_environment, strict: current_environment.nil?)
-                                .find_by(email: "#{username}".downcase)
+                                .find_by(email: "#{email}".downcase)
 
     unless user.present?
-      raise Keygen::Error::UnauthorizedError.new(
-        detail: 'email and password must be valid',
-        code: 'CREDENTIALS_INVALID',
-        header: 'Authorization',
-      )
+      raise Keygen::Error::InvalidCredentialsError.new(header: 'Authorization')
+    end
+
+    if user.single_sign_on_enabled?
+      redirect = sso_authorization_url_for(user)
+
+      raise Keygen::Error::SingleSignOnRequiredError.new(links: { redirect: })
     end
 
     if user.second_factor_enabled?
       otp = params.dig(:meta, :otp)
       if otp.nil?
-        raise Keygen::Error::UnauthorizedError.new(
-          detail: 'second factor is required',
-          code: 'OTP_REQUIRED',
-          pointer: '/meta/otp',
-        )
+        raise Keygen::Error::SecondFactorRequiredError.new(pointer: '/meta/otp')
       end
 
       unless user.verify_second_factor(otp)
-        raise Keygen::Error::UnauthorizedError.new(
-          detail: 'second factor must be valid',
-          code: 'OTP_INVALID',
-          pointer: '/meta/otp',
-        )
+        raise Keygen::Error::InvalidSecondFactorError.new(pointer: '/meta/otp')
       end
     end
 
     unless user.password? && user.authenticate(password)
```

Comment:
Should we return a `PASSWORD_REQUIRED` error code when SSO is not required, and the user has a password set but no password was provided? This could let us do the typical email first, password second, which allows SSO to replace password for some users (and eventually e.g. magic links, etc.)